segmentstore/server/src/test/java/io/pravega/segmentstore/server/attributes/AttributeIndexTests.java

```diff
@@ -672,18 +730,106 @@ public void testIndexCorruption() {
         context.index.cleanup(Collections.singleton(SEGMENT_ID));
     }
 
+    /**
+     * Tests the AttributeIndex recovery when the underlying Storage adapter supports atomic writes. It is expected that
+     * any metadata-stored Root Pointer is ignored in this case.
+     */
+    @Test
+    public void testRecoveryWithAtomicStorageWrites() {
+        val attribute = UUID.randomUUID();
+        val value1 = 1L;
+        val value2 = 2L;
+        val noRollingConfig = AttributeIndexConfig.builder().build(); // So we don't trip over rollovers.
+
+        @Cleanup
+        val context = new TestContext(noRollingConfig);
+        context.storage.supportsTruncation = false; // So we don't trip over truncations or rollovers.
+        context.storage.supportsAtomicWrites = true;
+        populateSegments(context);
+
+        // 1. Populate the index with first value
+        val idx = context.index.forSegment(SEGMENT_ID, TIMEOUT).join();
+        val rootPointer1 = idx.update(Collections.singletonMap(attribute, value1), TIMEOUT).join();
+        AssertExtensions.assertGreaterThan("Expected a valid Root pointer after update #2.", 0, rootPointer1);
+
+        val rootPointer2 = idx.update(Collections.singletonMap(attribute, value2), TIMEOUT).join();
+        AssertExtensions.assertGreaterThan("Expected a valid Root pointer after update #2.", rootPointer1, rootPointer2);
+
+        // 2. Update the Container Metadata with the first root pointer.
+        context.containerMetadata.getStreamSegmentMetadata(SEGMENT_ID)
+                .updateAttributes(Collections.singletonMap(Attributes.ATTRIBUTE_SEGMENT_ROOT_POINTER, rootPointer1));
+
+        // 3. Reload index and verify it ignores the metadata-stored root pointer (and uses the actual index).
+        context.index.cleanup(null);
+        val storageRead = new AtomicBoolean();
+        context.storage.readInterceptor = (name, offset, length, storage) -> CompletableFuture.runAsync(() -> storageRead.set(true));
+
+        val idx2 = context.index.forSegment(SEGMENT_ID, TIMEOUT).join();
+        val actualValue = idx2.get(Collections.singleton(attribute), TIMEOUT).join();
+        Assert.assertEquals("Unexpected value read.", value2, (long) actualValue.get(attribute));
+        Assert.assertTrue("Expecting storage reads after reload.", storageRead.get());
+    }
+
+    /**
+     * Tests the behavior of the Attribute Index when it is told that the underlying Storage adapter supports atomic
+     * writes but in fact it does not. It is expected to fail with corrupted data.
+     */
+    @Test
+    public void testRecoveryAfterIncompleteUpdateWithNoRootPointer() {
```

Comment:
I'm not sure I understood how we simulate atomic/non-atomic writes here.

Reply:
`   context.storage.supportsAtomicWrites = true;`

followed by lines 795-802 that add a partial write at the end of the segment.

This test verifies that the index gets corrupted if it trusts the Storage to say "it supports atomic writes" but in fact it does not.